os_dep.c

```diff
@@ -2058,28 +2074,87 @@ void GC_register_data_segments(void)
       return(result);
     }
 # else  /* !defined(__CHERI_PURE_CAPABILITY__) */
+    /* The CheriBSD LLVM compiler declares `etext`, `edata` and `end` as */
+    /* typeless variables. If bdwgc is statically linked with the main   */
+    /* executable, these capabilities are compiled with read-only        */
+    /* permissions and bounds that span the .data and .bss sections.     */
+    /*                                                                   */
+    /* If bdwgc is compiled as a shared library, these symbols are       */
+    /* compiled with zero bounds and cannot be dereferenced. Instead,    */
+    /* read-only capability returned by the loader is utilised           */
+    struct scan_bounds {
+        size_t start;
+        size_t end;
+        ptr_t  ld_cap;
+    };
+#   define SPANNING_CAPABILITY(cap, start, end) \
+      (cheri_base_get((cap)) <= (start)    \
+        && (cheri_base_get((cap)) + cheri_length_get((cap))) >= (end) \
+        && cheri_tag_get((cap)) \
+        && (cheri_perms_get((cap)) & (CHERI_PERM_LOAD|CHERI_PERM_LOAD_CAP)) != 0)
+
+    STATIC int GC_ld_cap_search(struct dl_phdr_info *info, size_t size, void *data)
+    {
+      struct scan_bounds *region = (struct scan_bounds *)data;
+      ptr_t load_addr = info->dlpi_addr;
+
+      if (!region || !SPANNING_CAPABILITY(load_addr,
+             region->start, region->end)) {
+        return 0;
+      }
+
+      region->ld_cap = cheri_bounds_set( cheri_address_set(load_addr
+                       , region->start), (size_t)(region->end - region->start));
```

Comment:
`, ` spilled off the previous line?

Reply:
changed in  [07b586b](https://github.com/capablevms/bdwgc/pull/19/commits/07b586b7ec011b520062977bd8e863a51061e711)